client/src/main/scala/org/http4s/client/websocket/WSClient.scala

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.client.websocket
+
+import cats._
+import cats.data.Chain
+import cats.data.OptionT
+import cats.effect._
+import cats.effect.concurrent.Deferred
+import cats.effect.concurrent.Ref
+import cats.effect.concurrent.TryableDeferred
+import cats.implicits._
+import fs2.Pipe
+import fs2.Stream
+import org.http4s.Headers
+import org.http4s.Method
+import org.http4s.Uri
+import scodec.bits.ByteVector
+
+/** A websocket request.
+  *
+  * @param uri
+  *   The URI.
+  * @param headers
+  *   The headers to send. Put your `Sec-Websocket-Protocol` headers here if needed. Some websocket
+  *   clients reject other WS-specific headers.
+  * @param method
+  *   The method of the intial HTTP request. Ignored by some clients.
+  */
+final case class WSRequest(
+    uri: Uri,
+    headers: Headers = Headers.empty,
+    method: Method = Method.GET,
+)
+
+sealed trait WSFrame extends Product with Serializable
+
+sealed trait WSControlFrame extends WSFrame
+
+sealed trait WSDataFrame extends WSFrame
+
+object WSFrame {
+  final case class Close(statusCode: Int, reason: String) extends WSControlFrame
+  final case class Ping(data: ByteVector) extends WSControlFrame
+  final case class Pong(data: ByteVector) extends WSControlFrame
+  final case class Text(data: String, last: Boolean = true) extends WSDataFrame
+  final case class Binary(data: ByteVector, last: Boolean = true) extends WSDataFrame
+}
+
+trait WSConnection[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Foldable, A <: WSFrame](wsfs: G[A]): F[Unit]
+
+  /** A `Pipe` which sends websocket frames and emits a `()` for each chunk sent. */
+  final def sendPipe: Pipe[F, WSFrame, Unit] = _.chunks.evalMap(sendMany(_))
```

Comment:
Nit: I find grouping the final methods after the abstract ones makes it easier to ready

Reply:
Like, all the way at the end of the class? Personally I prefer the proximity to `sendMany`.